File: apps/web/lib/services/searchService.ts
import { apiFetch } from "./apiClient";
import type {
  HubFeedResponse,
  SearchRequest,
  SearchResponse,
  SuggestResponse,
} from "@/lib/types/search";

const SEARCH_PREFIX = "/api/v1/search";

export const searchService = {
  search: ({ q, size = 20, tags, sort = "relevance", after }: SearchRequest) => {
    const params = new URLSearchParams();
    params.set("q", q);
    params.set("size", String(size));
    if (tags?.length) params.set("tags", tags.join(","));
    params.set("sort", sort);
    if (after) params.set("after", after);
    return apiFetch<SearchResponse>(`${SEARCH_PREFIX}?${params.toString()}`);
  },

  suggest: (prefix: string, size = 8) =>
    apiFetch<SuggestResponse>(
      `${SEARCH_PREFIX}/suggest?prefix=${encodeURIComponent(prefix)}&size=${size}`,
    ),

  hubFeed: (interestTags?: string[], page = 1, size = 8) => {
    const params = new URLSearchParams();
    if (interestTags?.length) params.set("interestTags", interestTags.join(","));
    params.set("page", String(page));
    params.set("size", String(size));
    const query = params.toString();
    return apiFetch<HubFeedResponse>(`${SEARCH_PREFIX}/hub-feed${query ? `?${query}` : ""}`);
  },
};

File: apps/server/src/main/java/com/chtholly/search/service/SearchService.java
package com.chtholly.search.service;

import com.chtholly.common.api.pagination.PageResponse;
import com.chtholly.post.api.dto.FeedItemResponse;
import com.chtholly.search.api.dto.HubFeedResponse;
import com.chtholly.search.api.dto.SuggestResponse;

import java.util.Collection;
import java.util.List;
import java.util.Map;

/**
 * 搜索服务接口：封装检索与联想能力，供控制器调用。
 */
public interface SearchService {
    /**
     * Searches published posts with optional tag filtering and cursor pagination.
     *
     * @param q query text
     * @param size maximum result count
     * @param tagsCsv optional comma-separated tag filter
     * @param after optional Base64URL cursor
     * @param sort requested result ordering
     * @param currentUserIdNullable optional current user ID
     * @return matching post page
     */
    PageResponse<FeedItemResponse> search(
            String q, int size, String tagsCsv, String after,
            SearchSort sort, Long currentUserIdNullable);

    /** Maps domain entity names to published articles using the structured analysis index. */
    PageResponse<FeedItemResponse> searchByEntityNames(
            List<String> entityNames, int size, Long currentUserIdNullable);

    /**
     * Aggregates Hub page search-backed regions with one msearch request.
     *
     * @param interestTags optional user interest tags for recommendations
     * @param currentUserIdNullable current user ID for liked/faved enrichment
     * @param page 1-indexed latest-posts page
     * @param size latest-posts page size
     */
    HubFeedResponse hubFeed(String interestTags, Long currentUserIdNullable, int page, int size);

    /**
     * 按兴趣标签权重推荐文章（function_score）。
     */
    List<FeedItemResponse> recommendByInterest(Map<String, Double> tagWeights,
                                               Collection<Long> excludePostIds,
                                               int limit,
                                               Long currentUserIdNullable);

    /**
     * 热门文章 fallback（按 like_count 排序）。
     */
    List<FeedItemResponse> recommendHot(Collection<Long> excludePostIds,
                                        int limit,
                                        Long currentUserIdNullable);

    /**
     * 基于种子帖子的标签/实体相似推荐。
     */
    List<FeedItemResponse> recommendSimilarToPost(long sourcePostId,
                                                  Collection<Long> excludePostIds,
                                                  int limit,
                                                  Long currentUserIdNullable);

    /**
     * 联想建议（Completion Suggester）。
     * @param prefix 前缀
     * @param size 返回条数
     */
    SuggestResponse suggest(String prefix, int size);
}
